src/parameterizations/vertical/MOM_geothermal.F90

```diff
@@ -304,6 +316,14 @@ subroutine geothermal(h, tv, dt, ea, eb, G, GV, CS, halo)
     enddo ; endif
   enddo ! j-loop
 
+! Calculate heat tendency due to addition and transfer of internal heat
+if (CS%id_internal_heat_tend_3d > 0) then
+  do k=1,nz ; do j=js,je ; do i=is,ie
+    work_3d(i,j,k) = GV%H_to_kg_m2 * tv%C_p * Idt * (h(i,j,k) * tv%T(i,j,k) - h_old(i,j,k) * temp_old(i,j,k))
```

Comment:
I'm thinking that allocating three new 3D arrays could become a problem.  Is it possible to get away with just using one array, say `work_3d`?  Something like this:
```
work_3d(i,j,k) =  h(i,j,k) * tv%T(i,j,k)
! ... more work
work_3d(i,j,) = GV%H_to_kg_m2 * tv%C_p * Idt * (h(i,j,k) * tv%T(i,j,k) - work_3d(i,j,k)
```
@adcroft  or anyone else, does that seem OK here?

Reply:
I don't see any problem with doing it like this.

Reply:
So what you're saying is to replace `h_old` and `T_old` and just store the multiplication of the two at the top of the file?

You will still need `h_old` because the diagnostic remapping has to post on the 'old' grid. At that point, you'll only be needing to save one 3D array. If you ever wanted to calculate a temperature tendency, you'd still need `T_old`. I think maybe just guarding this behind and `if` statement balances performance and utility?

Reply:
Yes, you are right, I missed the use of `h_old`.  And did not anticipate a future need for `T_old`.

You suggestion of `if (CS%id_internal_heat_tend_3d > 0)` is a good one, though the arrays will need to be changed to `allocatable`.

Reply:
Thanks both, I'll have a further go at this tomorrow.

Reply:
Would it make sense for me to code up diagnostics for heat, temperature and thickness all at once, in keeping with other diagnostics? In which case, allocating both h_old and T_old would be necessary but could be kept inside conditional statements depending on which diagnostics are called.

Reply:
I think it does since you've spent some time tracking down a missing 3D heat term. Maybe it'll save someone later on down the line from having to go through the same exercise.

Reply:
> Yes, you are right, I missed the use of `h_old`. And did not anticipate a future need for `T_old`.
> 
> Your suggestion of `if (CS%id_internal_heat_tend_3d > 0)` is a good one, though the arrays will need to be changed to `allocatable`.

Is the procedure for coding this up to specify `allocatable` when the variables are initially defined (i.e. `real, allocatable, dimension(...) :: h_old`, then use `allocate` within the conditional statement that @ashao suggests?